ChatQnA/Dockerfile

```diff
@@ -1,31 +1,51 @@
-
-
 # Copyright (C) 2024 Intel Corporation
 # SPDX-License-Identifier: Apache-2.0
 
-FROM python:3.11-slim
+ARG IMAGE_NAME=python
+ARG IMAGE_TAG=3.11-slim
+FROM ${IMAGE_NAME}:${IMAGE_TAG} as base
 
-RUN apt-get update -y && apt-get install -y --no-install-recommends --fix-missing \
+ENV LANG C.UTF-8
+
+RUN apt-get update -y && apt-get install -y \
     libgl1-mesa-glx \
-    libjemalloc-dev \
-    git
+    libjemalloc-dev && \
+    apt-get clean autoclean && \
+    apt-get autoremove -y && \
+    rm -rf /var/lib/apt/lists/*
 
 RUN useradd -m -s /bin/bash user && \
     mkdir -p /home/user && \
     chown -R user /home/user/
 
-WORKDIR /home/user/
-RUN git clone https://github.com/opea-project/GenAIComps.git
-
 WORKDIR /home/user/GenAIComps
+
+FROM base AS devel
+
+ARG GenAICompsRepo="https://github.com/opea-project/GenAIComps"
+ARG GenAICompsBranch="main"  # Branch name or Commit ID
+
+RUN apt-get update -y && apt-get install -y --no-install-recommends --fix-missing \
+    curl
+
+RUN curl -sSL --retry 5 ${GenAICompsRepo}/tarball/${GenAICompsBranch} | tar --strip-components=1 -xzf -
+
 RUN pip install --no-cache-dir --upgrade pip setuptools && \
     pip install --no-cache-dir -r /home/user/GenAIComps/requirements.txt && \
     pip install --no-cache-dir langchain_core
 
+FROM base AS prod
+
+ARG PYTHON=python3.11
+
 COPY ./chatqna.py /home/user/chatqna.py
 
 ENV PYTHONPATH=$PYTHONPATH:/home/user/GenAIComps
 
+COPY --from=devel /usr/local/lib/${PYTHON}/site-packages /usr/local/lib/${PYTHON}/site-packages
+COPY --from=devel /usr/local/bin /usr/local/bin
+COPY --from=devel /home/user/GenAIComps /home/user/GenAIComps
```

Comment:
When I tested using separate stage for ChatQnA `pip install`, and copying the files from there, like you do here, it saved only couple of MBs.  With shared base image, that's pretty insignificant, so I would skip it for simplicity.

Or are you seeing larger savings?

Reply:
Please double check your setup.

Reply:
I did a build with just pip install being done in separate step, or not, and image size difference reported by Docker was just few megs.  How did you test it, and what different did you see?

(I tested it with the comps base image, but that should not matter as it does exactly the same install step.)

Reply:
When caching is disabled, `pip` should clean out any downloading artefacts after installation, so I think only size difference would come from pip & setuptools upgrades, and those are only couple of MB in size. Which corresponds to what I saw.

Are you sure you're not confusing the size difference with disk space used for curl install?  `apt-get update` itself takes some space, and curl + its deps take ~4.5MB.

(I'm asking because if there **is** significant size difference, then I should consider same thing for base image.)